### ISSUE TITLE
ENH: Build line topology in dataframe for simple integration with NetworkX (+ potentially other graph systems)

### ISSUE THREAD

**Issue by iboates** [geopandas/geopandas#1592]
Hello there

I've been using geopandas for some time now, as well as The [NetworkX](https://github.com/networkx/networkx) library for doing graph operations. NetworkX has for a while now had their own `nx.from_shp` function for reading shapefiles and creating graphs out of them, but not much geospatial functionality beyond that. It is also based on raw GDAL which is apparently a bit of a pain for them to maintain dependencies for

They also have a function called `nx.from_pandas_edgelist` which can build a graph out of a pandas DataFrame so long as it has columns indicating the source and target of each edge. [I made a pull request to their repo with some code which can build a graph from a geopandas GeoDataFrame](https://github.com/networkx/networkx/pull/4107), so long as it only has LineString geometries. In the end, however, we realized that since `gpd.GeoDataFrmae` inherits from `pd.DataFrame`, it already technically compatible with `nx.from_pandas_edgelist`, it just needs to have the source and target columns pre-computed.

Some maintainers of NetworkX have agreed that shedding responsibility for handling geospatial I/O would be a relief to their project and I think that geopandas is in a good position to fill this void, given that it can already read & write from just about every GIS format and is already compatible with their support for pandas DataFrames

So I propose that I add a function to [geodataframe.py](https://github.com/geopandas/geopandas/blob/master/geopandas/geodataframe.py called `make_topology` which would construct `source` and `target` columns with unique ids where the nodes ought to be. Of course, the column names would be customizable and a spatial tolerance would be a parameter. Furthermore, it may also be interesting to have a way to somehow extract a point-based "nodes" GeoDataFrame for visualization of the identified nodes, but I'm not sure what the best way would be to do that. So the signature would look something like:

```python
gdf.make_topology(source_col="source", target_col="target", precision=0.005, inplace=False)
```

I didn't plan to do any topological cleaning or modification of geometry (i.e. assuming that the network is already clean), although maybe that would be an interesting set of new functions for the future.

I think this functionality would be quite useful for topological analysis of geographic networks in general, as the creation of the source & target columns would effectively make the GeoDataFrame ready for integration into NetworkX or indeed any other library or system which wants topological edge representations rather than geometric representations of features.

Here is the main bit of code which I had made for the original proposal for integration into NetworkX before we considered putting it in geopandas instead. It's not all that complex, basically we just look at all the LineString geometries first & last points and check to see if they are unique (within the specified precision)

```python

    # Just here for completeness in this example, since the original was designed for integration into NetworkX
    precision = 0.005
    source_col = "source"
    target_col = "target"

    # Determine number of dimensions of geometry by checking the first row (i.e. 2D or 3D lines?)
    dims = range(len(gdf[geometry][0].coords[0]))

    # Find all unique start & end points and assign them an id
    gdf["source_coords"] = gdf[geometry].apply(
        lambda geom: tuple(round(geom.coords[0][i], precision) for i in dims)
    )
    gdf["target_coords"] = gdf[geometry].apply(
        lambda geom: tuple(round(geom.coords[-1][i], precision) for i in dims)
    )
    node_ids = {}
    i = 0
    for row in gdf.itertuples(index=False):
        node_1 = row.source_coords
        node_2 = row.target_coords
        if node_1 not in node_ids:
            node_ids[node_1] = i
            i += 1
        if node_2 not in node_ids:
            node_ids[node_2] = i
            i += 1

    # Assign the unique id to each
    gdf[source_col] = gdf["source_coords"].apply(lambda x: node_ids[x])
    gdf[target_col] = gdf["target_coords"].apply(lambda x: node_ids[x])
```

**Comment by brendan-ward** [comment]
@iboates Thanks for sharing your ideas on this!

Provided I follow your example, it looks like the output is a GeoDataFrame attributed with source and target columns that can then be used as input to NetworkX?  As such, it doesn't look like you are suggesting adding NetworkX as an optional dependency to GeoPandas?

A few thoughts:
1. we should be careful with the term "topology" as it could easily be conflated between graph topology (as used here) and spatial formats / representations of topology (e.g., TopoJSON) that are inputs to other operations (e.g., #1387).  Over the long term, there may be value to handling both types in GeoPandas.  Not saying we shouldn't use the term, but that we need to make sure it is unambiguous in specific functionality.

2. it seems like the ideas here should sit on top of spatial joins or at least leverage the underlying spatial index.  Identifying shared endpoints between line segments is a very specific type of topology, but there could be a lot of value to generalizing that further.  For example, in my work, I've been using GeoPandas + NetworkX to find all sets of adjacent geometries (e.g., polygons) in order to dissolve those that touch into a single feature.  This is basically a spatial join of geometries against themselves, dropping the self-self joins, and then passing to NetworkX as edges in order to identify all connected sets of geometries.  I've also been using a similar approach to find "neighborhoods" of features where individual features are within a tolerance of each other.  This could be implemented as a spatial join using some nearest-neighbor tolerance, which is already under discussion in GeoPandas (e.g., #1271, #1096), or implemented as a spatial join against buffers of the original geometries.

Or put a bit differently: this generalizes to a) an operation (e.g., spatial join, nearest, etc) which creates pairs of records in a data frame, b) input to NetworkX (outside GeoPandas?), c) user does whatever network / graph operations they need using NetworkX directly.  Thus "topology" in this case is just the interpretation of those pairs, not a prescription of what those pairs must be.

3. since the potential use cases under 2 are highly variable, I wonder if this would be better handled as a series of examples on how to use GeoPandas + NetworkX, rather than adding functionality directly to GeoPandas?  That said, it may be useful to expose constructs in sjoin / sindex that make interoperability with NetworkX easier; would need to give that some thought to see what APIs are needed.

4. given the above, I wonder if it might be better to instead focus on a function that addresses your specific use case here, which is neither a spatial join nor precisely a nearest neighbor join.  Something that takes source LineStrings and target LineStrings (could be source) as inputs and outputs all pairs where the endpoints are within some tolerance.  Internally this would extract the endpoints and do a tolerance-based join.

**Comment by martinfleis** [comment]
Just quickly (I'll add longer version later), I have workflow geopandas-networkx implemented in momepy using libpysal, which does topology in several ways (Rook, Queen, fuzzy). lipbysal.Weights object can be directly exported to networkx.Graph. So these functions already exist in the ecosystem and are well optimised. What I'd do in GeoPandas would be adding an example using these.

Btw, the best example of this is `mapclassify.greedy` module, which first generates topology using libysal and then can use networkx algorithms to get topological coloring. See https://github.com/pysal/mapclassify/blob/master/mapclassify/greedy.py.

**Comment by jarrodmillman** [comment]
I am a networkx core developer and I wanted to add a bit about our perspective.  As mentioned, we currently have two functions to handle shapefiles:  ``read_shp`` and ``write_shp``.  We have one docstring example of how to use it:
```
    >>> G = nx.read_shp("test.shp")  # doctest: +SKIP
```
The tests check that we can read what we write, but they don't directly check that we can read valid shapefiles or whether the shapefiles we write are valid.  I suspect someone would tell us if they didn't work, but that isn't very reassuring.

None of the current networkx core developers know anything about shapefiles or geospatial data.  (I can't follow much of the above discussion as a result.)

Moreover, GDAL is our most difficult dependency.  I spend more time dealing with it than almost anything else involving our CI system.  Users complain that they have difficulty installing it.  It is an optional dependency, but many users want to install all dependencies for whatever reason.  As a result, I've been considering making it more difficult for users to discover that it is a dependency.

I was excited when @iboates made a networkx PR (https://github.com/networkx/networkx/pull/4107) to support geopandas because I hoped it might provide a way forward for us to better support geospatial researchers without having to deal with GDAL.  Infortunately, since no one on our developer team knows much about spatial data we can't realistically even thoroughly review the PR.

We are planning a new major release of networkx and I want to use this opportunity to remove ``read_shp`` and ``write_shp``.  But I want to make sure that this doesn't cause a problem for users.  At a minimum, I was imaging that we could add several gallery/tutorial style examples of using nx with geospatial data.  Of course, we would need help with that b/c we don't know how to work with geospatial data.  I wasn't even sure how often folks use networkx with geospatial data.

I suspect one of us will need to depend on the other.  NetworkX is pure python and the only real difficulty installing it is getting GDAL installed if you want that optional functionality.  So I think it would be easy for geopandas to include networkx as a dependency.  We could include geopandas as dependency of networkx, but then we still have a (slightly more hidden) dependency on GDAL, which I would like to avoid.  If you decide to make networkx a dependency, we would be happy to work with you on any questions you have.

I am not sure how to proceed.  I suspect folks here will have more ideas about this than I will.

In summary, we (the nx developers) would like to remove ``read_shp`` and ``write_shp`` in our next major release.  We would like to have an answer for our ``read_shp`` and ``write_shp`` users about what they should do.   If we need to keep ``read_shp`` and ``write_shp``, we need help adding tests and gallery examples about how to use them.  Ideally, we can remove `read_shp`` and ``write_shp`` and point users to a gallery of examples of using geospatial data and networkx together.  I would like some of those examples to live in our gallery of examples (so we can test that we don't break anything), but it probably makes sense for us to also provide some links to gallery examples in other projects.

**Comment by dschult** [comment]
Hello!,    Like @jarrodmillman I too am a NetworkX core developer and am interested in making a smooth interface between geopandas and networkx that requires minimal maintenance.  If I understand the comments written above, @iboates approach would allow GeoPandas and NetworkX to coexist with an easy interface for communication. It would also allow us both to avoid an explicit dependency on the other.  That is, GeoPandas would have a function that adds columns to the underlying DataFrame for source and target nodes of each edge.  That function would not require NetworkX at all.  NetworkX would have a function that reads the network structure from any DataFrame that has a column representing source nodes and a column representing target nodes for each edge. That function would not depend on GeoPandas.

We could both have examples that use the other package.  I suppose that means we'd both need CI tools that install the other package. But only people interested in building the examples would need both. NX wouldn't be needed to install GeoPandas and GeoPandas wouldn't be needed to install NetworkX.  We would be willing to contribute to this effort if there is interest in providing this bridge between packages along with examples to show the way.

As for the name "Topology", perhaps something like "network_structure" would be more specific than topology and avoid confusion.

**Comment by rossbar** [comment]
Another quick 2 cents from an NX core contributor --- A few high-level considerations stand out to me as being the most important: 1) making sure that users are using the right tool for the job and 2) that users don't lose any important functionality. Having never worked with geospatial data myself, it does seem that a package like GeoPandas is a much more appropriate place for users to expect to find things like low-level I/O with geospatial data formats. I especially agree with @brendan-ward and @jarrodmillman 's points about having extensive examples to clearly illustrate how NetworkX and GeoPandas can be used in concert to perform geospatial data analysis.

**Comment by martinfleis** [comment]
I think we have to distinguish between the two use cases. 

1. Generating networkx.Graph from LineString geometry representing pseudo-planar street network (and generating GeoDataFrame from Graph)
2. The general interface between geospatial data and networkx.

The **first** case is somehow implemented as `networkx.read_shp` and `write_shp` and I fully understand the need to get rid of GDAL dependency (in fact, we're trying to do the same). Moreover, such a function should be more flexible than `shp` input. I believe (but I might be wrong), that what we're talking about is implemented in [momepy](https://github.com/martinfleis/momepy) as [`gdf_to_nx`](http://docs.momepy.org/en/stable/generated/momepy.gdf_to_nx.html#momepy.gdf_to_nx) and [`nx_to_gdf`](http://docs.momepy.org/en/stable/generated/momepy.nx_to_gdf.html#momepy.nx_to_gdf). It may not support everything you are looking for at the moment (e.g. DiGraph), but it is fairly stable now. 

The question here is where would be the right place for this. I have it in momepy as it does quite specific things (not mentioning that it can generate dual graph), so I did not feel that it should go directly to GeoPandas. And as a GeoPandas dev, I'm still not sure if it should be in here. To make it work you need very specific data and that is something we're generally trying to avoid. If there's a function, it should ideally work on all geometry types. Having it in networkx makes sense to me, but that would require optionally depend on GeoPandas, which can at the moment be a bit tricky. Not so much in the future, when only C dependency could be shapely 2.0 (no need for rtree anymore and GDAL could be optional at that moment), but no one knows when that is going to happen to be honest.

Keeping it _only_ in momepy makes it fairly invisible as momepy is quite narrowly focused. What about OSMnx? That is already an established package focusing exactly on this, depending on GeoPandas and networkx already. That could become a natural home (cc @gboeing) for such functionality.

The **second** case is more complex and covers to a degree the first one plus general description of topology. In geospatial world, that is normally captured by spatial weights matrices, which are currently well handled by `libpysal` and I honestly don't see the need to have an explicit function `gdf_to_graph`, since that would be duplicating libpysal's functionality. libpysal.Weights object has the interface to networkx already built-in. The greedy module in mapclassify I mentioned above is the prime example of this second case. It starts from GeoDataFrame of any geometry type, builds spatial weights matrix in libpysal and does the analysis on its networkx version.

I think we need to clarify here what are we talking about and what should be the general goal. My understanding is that the first case is what networkx dev team wants and in that case, the crucial question here is _what is the right home_. Then we can move to details (moving existing functions around, merging with other...). What I would like to avoid is having multiple incompatible implementations in various packages around.

**Comment by gboeing** [comment]
It may be of interest how OSMnx handles NetworkX MultiDiGraph to/from Geopandas GeoDataFrame conversions with its [graph_to_gdfs](https://github.com/gboeing/osmnx/blob/f05b359c418ff0139f6696d504b766a35775ef7c/osmnx/utils_graph.py#L17) and [graph_from_gdfs](https://github.com/gboeing/osmnx/blob/f05b359c418ff0139f6696d504b766a35775ef7c/osmnx/utils_graph.py#L104) functions.

The package has undergone a major renovation over the summer and has shifted somewhat away from its original mission of being a pure OpenStreetMap -> NetworkX graph utility, towards more flexibly working with OSM spatial networks in NetworkX and OSM non-networked geometries in Geopandas (particularly with the new [geometries](https://github.com/gboeing/osmnx/pull/549) module).

**Comment by ljwolf** [comment]
Building topologies from geographic data stored in `GeoDataFrames` is indeed a big interest to many. As @martinfleis mentions above, the [`pysal` library](https://geographicdata.science/book/notebooks/04_spatial_weights.html#Spatial-Weights-from-real-world-data) has also focused on providing these kinds of fast algorithms for graph/topology construction with no GEOS dependency. Though, in econometrics, these are called "spatial weights," rather than "graphs" for historical reasons. We cover many kinds of distance-based and contiguity-based weights. I asked about that [a while ago](https://github.com/networkx/networkx/issues/3067), and that stalled. 

So, we went ahead and did it in `libpysal`. We have converters `to_networkx()` and `from_networkx()`, as well as `to_adjlist()` and `from_adjlist()` and access to the underlying sparse matrix representation 

But... it'd be great if we could have a single library where this kind of graph building is fast and performant... ours [usually is fast](http://ljwolf.org/notebooks/perftesting_graphs.html) because "sharing a vertex" is slightly different from "touches at a point," but building on top of pygeos might open new performance benefits.

Regardless, coordinating to build a single topology builder that has a generic builder for things with a `__geo_interface__` and possibly has faster implementations for specific input types would be really great, and I think consolidate a ton of duplicated effort in the ecosystem. 

For those not familiar with `pysal`, an example task might be:
```python
import geopandas, osmnx
from pysal.lib import weights

graph = osmnx.graph_from_place("Bristol, UK")
nodes, edges = osmnx.graph_to_gdfs(graph)
```
Edges is now a geodataframe of line segments, as if read in from file. In our lexicon, we'd use "queen contiguity" to refer to neighboring geometries that share a single vertex: 
```python
w = weights.Queen.from_dataframe(edges)
w.to_adjlist() # contains head id, tail id, link weight

w.to_networkx() # contains the NetworkX Graph/DiGraph

w.sparse # contains the scipy.sparse adjacency matrix
```
This works for polygons, too: 
```python
boros = geopandas.read_file(geopandas.datasets.get_path("nybb"))
weights.Rook.from_dataframe(boros)
```
and, for distance-based graphs, we have a similar API:
```python
weights.DistanceBand.from_dataframe(nodes, threshold=.1)
weights.KNN.from_dataframe(nodes, k=10)
weights.Kernel.from_dataframe(nodes, function='gaussian')
```

**Comment by jarrodmillman** [comment]
NetworkX is deprecating read/write functionality for shapefiles (#4298) and providing gallery examples for working with geospatial data using the main packages for geospatial data analysis for Python (#4366).  We are interested in making sure our gallery examples provide the recommended way to read/write geospatial data and converting them back and forth to NetworkX.  We would love feedback and additional gallery examples!  One of our main goals is to make it clear to our users how to get started and where to go for questions regarding geospatial data formats since we (the NetworkX developers) don't work with geospatial data and don't know how to provide the correct guidance for users.

We think this is the best way forward and hope this will lead to more cross-project discussions.  If you have questions about NetworkX, please don't hesitate to @mention us on related issues PRs here.
